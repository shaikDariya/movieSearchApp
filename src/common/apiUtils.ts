export const apiBaseUrl: string = "https://www.omdbapi.com/?apikey=eaa9d1e0";
const jsonWrapper = (method: string) => async (
  path: string,
  init: any = {}
): Promise<any> => {
  const response = await fetch(`${apiBaseUrl}&${path}`, {
    ...init,
    method,
    headers: {
      ...init.headers
    }
  });
  if (!response.ok) {
    throw new Error(response.status.toString());
  }
  try {
    return await response.json();
  } catch (e) {
    console.warn("Json Parse Error", e);
    return null;
  }
};

export const fetchJson = jsonWrapper("get");
