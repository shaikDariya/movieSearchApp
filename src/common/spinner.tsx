import React from "react";
import Spinner from "react-bootstrap/Spinner";
const SpinnerContainer = () => (
  <div className="row align-items-center justify-content-center mt-6">
    <Spinner animation="border">
      <span className="sr-only">Loading...</span>
    </Spinner>
  </div>
);
export default SpinnerContainer;
