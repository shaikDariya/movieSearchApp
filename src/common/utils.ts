export const sortByField = (list: any[], field: string): any[] => {
  return list.slice().sort((f: any, s: any) => {
    const fField = f[field];
    const sField = s[field];
    if (fField > sField) {
      return 1;
    }

    if (fField < sField) {
      return -1;
    }
    return 0;
  });
};
