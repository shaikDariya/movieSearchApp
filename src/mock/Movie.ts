export const mockMovie = {
  Search: [
    {
      Title: "testing 1",
      Year: "2011–2013",
      imdbID: "tt2959006",
      Type: "series",
      Poster:
        "https://m.media-amazon.com/images/M/MV5BYjdlNzNiYzEtMDRkOC00ZWU0LTk5N2ItMzc5MThiZDYyYmFjXkEyXkFqcGdeQXVyNjMxNzQ2NTQ@._V1_SX300.jpg"
    },
    {
      Title: "testing 2",
      Year: "2010",
      imdbID: "tt1863391",
      Type: "movie",
      Poster: "N/A"
    },
    {
      Title: "testing 3",
      Year: "2015",
      imdbID: "tt4957280",
      Type: "movie",
      Poster: "N/A"
    },
    {
      Title: "testing 4",
      Year: "2005–2006",
      imdbID: "tt5028506",
      Type: "series",
      Poster: "N/A"
    },
    {
      Title: "testing 5",
      Year: "2012",
      imdbID: "tt3014410",
      Type: "movie",
      Poster: "N/A"
    },
    {
      Title: "testing 6",
      Year: "2014",
      imdbID: "tt3844108",
      Type: "movie",
      Poster: "N/A"
    }
  ],
  totalResults: "6",
  Response: "True"
};

export const mockMovieDetail = {
  Title: "Pon un hombre en tu vida",
  Year: "1996",
  Rated: "PG-13",
  Released: "23 Aug 1996",
  Runtime: "87 min",
  Genre: "Comedy",
  Director: "Eva Lesmes",
  Writer: "Eva Lesmes, Luis Marías, Joaquín Oristrell",
  Actors: "Toni Cantó, Cristina Marcos, Javier Cámara, Anabel Alonso",
  Plot:
    "In a swimming pool accident, Juan and Belinda somehow exchange bodies. Juan is coaching a football team for an important match and Belinda is getting married. They agree to co-operate and continue in their new bodies.",
  Language: "Spanish",
  Country: "Spain",
  Awards: "N/A",
  Poster:
    "https://m.media-amazon.com/images/M/MV5BMTkzMTUxNjQ3MV5BMl5BanBnXkFtZTcwNTQwNDMzMQ@@._V1_SX300.jpg",
  Ratings: [
    {
      Source: "Internet Movie Database",
      Value: "6.2/10"
    }
  ],
  Metascore: "N/A",
  imdbRating: "6.2",
  imdbVotes: "171",
  imdbID: "tt0117358",
  Type: "movie",
  DVD: "N/A",
  BoxOffice: "N/A",
  Production: "N/A",
  Website: "N/A",
  Response: "True"
};
