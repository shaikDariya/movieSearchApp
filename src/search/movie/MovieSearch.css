.img-icon{
    height: 24px;
    margin-right: 10px;
    width: 24px;
}