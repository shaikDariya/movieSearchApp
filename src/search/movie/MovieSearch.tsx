import React, { useReducer } from "react";
import { AsyncTypeahead } from "react-bootstrap-typeahead";
import MovieReducer, { movieState } from "./MovieReducer";

import "react-bootstrap-typeahead/css/Typeahead.css";
import "./MovieSearch.css";
import { dispatchSetLoading, dispatchFetchMovie } from "./MovieAction";
import { useHistory } from "react-router-dom";
import { MovieSuggestionType } from "./MovieType";
import { fetchJson } from "../../common/apiUtils";

export const nAImageLink =
  "https://m.media-amazon.com/images/G/01/imdb/images/nopicture/32x44/film-3119741174._CB468665901_.png";

const MovieSearch = () => {
  const [state, dispatch] = useReducer(MovieReducer, movieState);
  const history = useHistory();

  const handleSearch = async (query: string) => {
    dispatch(dispatchSetLoading(true));
    const { Search } = await fetchJson(`s=${query}`);
    dispatch(dispatchFetchMovie(Search));
    dispatch(dispatchSetLoading(false));
  };
  const handleSelect = (selected: any) => {
    const { imdbID } = selected.pop();
    history.push(`/movie/${imdbID}`);
  };
  return (
    <div className="d-flex justify-content-center mt-2">
      <div className="col-8 col-sm-6 col-md-4">
        <AsyncTypeahead
          id="movie-search"
          {...state}
          minLength={3}
          clearButton
          labelKey="Title"
          onSearch={handleSearch}
          onChange={handleSelect}
          placeholder="Search for a Movie ..."
          renderMenuItemChildren={option => <MovieTitle {...option} />}
        />
      </div>
    </div>
  );
};

const MovieTitle = ({ Title, imdbID, Poster }: MovieSuggestionType) => {
  return (
    <div>
      <img
        alt="ass"
        src={Poster === "N/A" ? nAImageLink : Poster}
        className="img-icon"
      />
      <span>{Title}</span>
    </div>
  );
};

export default MovieSearch;
