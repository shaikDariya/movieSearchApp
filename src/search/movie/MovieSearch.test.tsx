import React from "react";
import { Router } from "react-router";
import { createMemoryHistory } from "history";
import {
  render,
  cleanup,
  waitForDomChange,
  fireEvent
} from "@testing-library/react";
import "@testing-library/jest-dom";
import fetchMock from "fetch-mock";

import MovieSearch from "./MovieSearch";
import { mockMovie } from "../../mock/Movie";
import { apiBaseUrl } from "../../common/apiUtils";
import { act } from "react-dom/test-utils";

jest.mock("popper.js", () => {
  const PopperJS = jest.requireActual("popper.js");

  return class {
    static placements = PopperJS.placements;

    constructor() {
      return {
        destroy: () => {},
        scheduleUpdate: () => {}
      };
    }
  };
});

const history = createMemoryHistory();
const renderMovieSearch = () => {
  return render(
    <Router history={history}>
      <MovieSearch />
    </Router>
  );
};

describe("MovieSearch", () => {
  afterEach(() => {
    fetchMock.reset();
    cleanup();
  });
  it("MovieSearch renderer", () => {
    const { asFragment } = renderMovieSearch();
    expect(asFragment()).toMatchSnapshot();
  });
  it("Input with Search for a Movie ...", () => {
    const { getByPlaceholderText } = renderMovieSearch();
    const inputElment = getByPlaceholderText(/Search for a Movie .../);
    expect(inputElment).toBeInTheDocument();
  });
  describe("Search", () => {
    it("with below 3 keywords should not show suggestion", () => {
      const { container, getByPlaceholderText } = renderMovieSearch();
      const inputElment = getByPlaceholderText(/Search for a Movie .../);
      fireEvent.change(inputElment, { target: { value: "ab" } });
      expect(container.querySelector(".dropdown-menu")).not.toBeInTheDocument();
    });
    it("with 3 keywords should show suggestion", async () => {
      const { container, getByPlaceholderText } = renderMovieSearch();
      const inputElment = getByPlaceholderText(/Search for a Movie .../);
      act(() => {
        fireEvent.change(inputElment, { target: { value: "abc" } });
      });
      await waitForDomChange();
      expect(container.querySelector(".dropdown-menu")).toBeInTheDocument();
    });
    it("Mock movie reults should shown in suggestion dropdown", async () => {
      const query = "test";
      fetchMock.getOnce(`${apiBaseUrl}&s=${query}`, mockMovie);
      const { container, getByPlaceholderText } = renderMovieSearch();
      const inputElment = getByPlaceholderText(/Search for a Movie .../);
      act(() => {
        fireEvent.change(inputElment, { target: { value: query } });
      });
      await waitForDomChange();
      const findDropDownItems = container.querySelectorAll(
        ".dropdown-menu .dropdown-item"
      );
      expect(findDropDownItems.length).toEqual(mockMovie.Search.length);
    });
    it("Suggestion DropDown onSelect Should Navigate to Details Page", async () => {
      const query = "test";
      fetchMock.getOnce(`${apiBaseUrl}&s=${query}`, mockMovie);
      const { container, getByPlaceholderText } = renderMovieSearch();
      const inputElment = getByPlaceholderText(/Search for a Movie .../);
      act(() => {
        fireEvent.change(inputElment, { target: { value: query } });
      });
      await waitForDomChange();
      const findDropDownItems = container.querySelectorAll(
        ".dropdown-menu .dropdown-item"
      );
      act(() => {
        fireEvent.click(findDropDownItems[0]);
      });
      expect(history.location.pathname).toEqual(
        `/movie/${mockMovie.Search[0].imdbID}`
      );
    });
  });
});
