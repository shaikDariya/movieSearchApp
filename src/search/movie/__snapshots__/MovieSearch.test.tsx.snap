// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieSearch MovieSearch renderer 1`] = `
<DocumentFragment>
  <div
    class="d-flex justify-content-center mt-2"
  >
    <div
      class="col-8 col-sm-6 col-md-4"
    >
      <div
        class="rbt"
        style="outline: none; position: relative;"
        tabindex="-1"
      >
        <div
          style="display: flex; flex: 1; height: 100%; position: relative;"
        >
          <input
            aria-autocomplete="both"
            aria-expanded="false"
            aria-haspopup="listbox"
            autocomplete="off"
            class="rbt-input-main form-control rbt-input"
            placeholder="Search for a Movie ..."
            role="combobox"
            type="text"
            value=""
          />
          <input
            aria-hidden="true"
            class="rbt-input-hint"
            readonly=""
            style="background-color: transparent; border-color: transparent; box-shadow: none; color: rgba(0, 0, 0, 0.35); left: 0px; pointer-events: none; position: absolute; top: 0px; width: 100%;"
            tabindex="-1"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
